## **Práctica 2 - Generación de Variables Aleatorias Discretas**

### **Ejercicio 1**

In [1]:
import pandas as pd

demanda = [0, 1, 2, 3, 4, 5]
probabilidad = [0.05, 0.15, 0.30, 0.30, 0.15, 0.05]

tabla_distribucion = pd.DataFrame({
    "Demanda diaria": demanda,
    "Probabilidad": probabilidad
})

tabla_distribucion["Probabilidad acumulada"] = tabla_distribucion["Probabilidad"].cumsum()

tabla_distribucion

,Demanda diaria,Probabilidad,Probabilidad acumulada
0,0,0.05,0.05
1,1,0.15,0.20
2,2,0.30,0.50
3,3,0.30,0.80
4,4,0.15,0.95
5,5,0.05,1.00


In [2]:
def obtener_demanda(numero_aleatorio):
    for i in range(len(tabla_distribucion)):
        if numero_aleatorio <= tabla_distribucion.loc[i, "Probabilidad acumulada"]:
            return tabla_distribucion.loc[i, "Demanda diaria"]

obtener_demanda(0.67)

np.int64(3)

In [3]:
import random

random.seed(7)

dias = 15
inventario_diario = 3
resultados = []

for dia in range(1, dias + 1):
    numero_aleatorio = random.random()
    demanda_simulada = obtener_demanda(numero_aleatorio)
    ventas = min(demanda_simulada, inventario_diario)
    faltante = max(demanda_simulada - inventario_diario, 0)
    sobrante = max(inventario_diario - demanda_simulada, 0)

    if faltante > 0:
        resultado = f"Faltaron {faltante} botella(s)"
    elif sobrante > 0:
        resultado = f"Sobraron {sobrante} botella(s)"
    else:
        resultado = "Se vendió todo"

    resultados.append({
        "Día": dia,
        "Número aleatorio": round(numero_aleatorio, 4),
        "Demanda simulada": demanda_simulada,
        "Ventas": ventas,
        "Faltante": faltante,
        "Sobrante": sobrante,
        "Resultado": resultado
    })

tabla_simulacion = pd.DataFrame(resultados)
tabla_simulacion

,Día,Número aleatorio,Demanda simulada,Ventas,Faltante,Sobrante,Resultado
0,1,0.3238,2,2,0,1,Sobraron 1 botella(s)
1,2,0.1508,1,1,0,2,Sobraron 2 botella(s)
2,3,0.6509,3,3,0,0,Se vendió todo
3,4,0.0724,1,1,0,2,Sobraron 2 botella(s)
4,5,0.5359,3,3,0,0,Se vendió todo
5,6,0.3657,2,2,0,1,Sobraron 1 botella(s)
6,7,0.0580,1,1,0,2,Sobraron 2 botella(s)
7,8,0.5074,3,3,0,0,Se vendió todo
8,9,0.0375,0,0,0,3,Sobraron 3 botella(s)
9,10,0.4336,2,2,0,1,Sobraron 1 botella(s)


In [4]:
resumen = pd.DataFrame({
    "Métrica": [
        "Demanda total",
        "Ventas totales",
        "Botellas faltantes",
        "Botellas sobrantes",
        "Días con faltante",
        "Días con sobrante",
        "Promedio de demanda diaria"
    ],
    "Valor": [
        tabla_simulacion["Demanda simulada"].sum(),
        tabla_simulacion["Ventas"].sum(),
        tabla_simulacion["Faltante"].sum(),
        tabla_simulacion["Sobrante"].sum(),
        (tabla_simulacion["Faltante"] > 0).sum(),
        (tabla_simulacion["Sobrante"] > 0).sum(),
        round(tabla_simulacion["Demanda simulada"].mean(), 2)
    ]
})

resumen

,Métrica,Valor
0,Demanda total,27.0
1,Ventas totales,26.0
2,Botellas faltantes,1.0
3,Botellas sobrantes,19.0
4,Días con faltante,1.0
5,Días con sobrante,11.0
6,Promedio de demanda diaria,1.8


#### Conclusión

Con los resultados de la simulación, mantener 3 botellas diarias es una decisión aceptable, ya que solamente hubo 1 botella faltante en los 15 días. Sin embargo, también sobraron 19 botellas en total y en 11 días quedó inventario sin vender. Por eso, aunque la política evita casi todas las pérdidas por falta de producto, también genera bastante producto sobrante. En este caso, consideraría mantener las 3 botellas si se quiere evitar perder ventas, pero si el objetivo fuera reducir inventario sobrante, podría evaluarse bajar a 2 botellas algunos días.

### **Ejercicio 2**

Para generar la variable aleatoria X se usa el método de la probabilidad acumulada. Primero se divide el intervalo de 0 a 1 según las probabilidades dadas. Luego se genera un número aleatorio U y, dependiendo del intervalo donde caiga, se le va a asignar el valor correspondiente de X.



In [5]:
valores_x = [1, 2, 3, 4]
probabilidades_x = [0.30, 0.20, 0.35, 0.15]

tabla_x = pd.DataFrame({
    "X": valores_x,
    "Probabilidad": probabilidades_x
})

tabla_x["Probabilidad acumulada"] = tabla_x["Probabilidad"].cumsum()

tabla_x

,X,Probabilidad,Probabilidad acumulada
0,1,0.30,0.30
1,2,0.20,0.50
2,3,0.35,0.85
3,4,0.15,1.00


In [8]:
def generar_x():
    u = random.random()

    if u <= 0.30:
        x = 1
    elif u <= 0.50:
        x = 2
    elif u <= 0.85:
        x = 3
    else:
        x = 4

    return u, x

generar_x()

(0.9477089424570057, 4)

### **Ejercicio 3**



In [9]:
random.seed(12)

n_cartas = 100
simulaciones = 10000
cartas = list(range(1, n_cartas + 1))
aciertos = []

for _ in range(simulaciones):
    baraja = cartas.copy()
    random.shuffle(baraja)

    total_aciertos = 0
    for posicion, carta in enumerate(baraja, start=1):
        if posicion == carta:
            total_aciertos += 1

    aciertos.append(total_aciertos)

esperanza_estimada = sum(aciertos) / simulaciones
varianza_estimada = pd.Series(aciertos).var(ddof=0)

comparacion = pd.DataFrame({
    "Medida": ["Esperanza", "Varianza"],
    "Estimación por simulación": [round(esperanza_estimada, 4), round(varianza_estimada, 4)],
    "Valor exacto": [1, 1]
})

comparacion

,Medida,Estimación por simulación,Valor exacto
0,Esperanza,0.9859,1
1,Varianza,1.0001,1


#### Conclusión

Los valores estimados por simulación quedan cercanos a los valores exactos. En este caso, tanto la esperanza como la varianza exacta del número de aciertos son 1. La diferencia que pueda aparecer se debe a que la simulación usa una cantidad finita de repeticiones, pero al aumentar el número de simulaciones los resultados deberían acercarse más al valor teórico.

### **Ejercicio 4**

Como p = 0.8, es más probable obtener éxitos que fracasos. Por eso, en vez de generar una variable Bernoulli una por una, se puede usar un método más eficiente que calcula cuántos éxitos aparecen antes del próximo fracaso.

Para eso se usa la fórmula:

$$N = Int\left(\frac{\ln(U)}{\ln(p)}\right) + 1$$

El valor de N indica en qué posición aparece el siguiente fracaso. Entonces se colocan N - 1 éxitos, y luego un fracaso. Y se debe de repetir el proceso hasta completar las 25 variables.

In [11]:
import math

random.seed(123)

p = 0.8
n = 25
posicion = 1
numeros_usados = 0
valores_bernoulli = []
pasos = []

while posicion <= n:
    u = random.random()
    numeros_usados += 1
    N = int(math.log(u) / math.log(p)) + 1

    inicio = posicion
    exitos = min(N - 1, n - posicion + 1)

    for _ in range(exitos):
        valores_bernoulli.append(1)
        posicion += 1

    if posicion <= n:
        valores_bernoulli.append(0)
        fin = posicion
        posicion += 1
        fracaso = "Sí"
    else:
        fin = n
        fracaso = "No, queda fuera"

    pasos.append({
        "Paso": numeros_usados,
        "U": round(u, 4),
        "N": N,
        "Desde posición": inicio,
        "Hasta posición": fin,
        "Éxitos antes del fracaso": exitos,
        "Fracaso incluido": fracaso
    })

tabla_pasos = pd.DataFrame(pasos)
tabla_pasos

,Paso,U,N,Desde posición,Hasta posición,Éxitos antes del fracaso,Fracaso incluido
0,1,0.0524,14,1,14,13,Sí
1,2,0.0872,11,15,25,10,Sí


In [12]:
tabla_bernoulli = pd.DataFrame({
    "Variable": range(1, n + 1),
    "Resultado": valores_bernoulli
})

resumen_bernoulli = pd.DataFrame({
    "Métrica": [
        "Cantidad de variables generadas",
        "Cantidad de unos",
        "Cantidad de ceros",
        "Números aleatorios utilizados"
    ],
    "Valor": [
        n,
        tabla_bernoulli["Resultado"].sum(),
        n - tabla_bernoulli["Resultado"].sum(),
        numeros_usados
    ]
})

display(tabla_bernoulli)
resumen_bernoulli

,Variable,Resultado
0,1,1
1,2,1
2,3,1
3,4,1
4,5,1
5,6,1
6,7,1
7,8,1
8,9,1
9,10,1


,Métrica,Valor
0,Cantidad de variables generadas,25
1,Cantidad de unos,23
2,Cantidad de ceros,2
3,Números aleatorios utilizados,2



Con el método eficiente se necesitaron 2 números aleatorios para generar las 25 variables Bernoulli. Esto pasa porque cada número aleatorio permite generar un bloque de éxitos seguidos hasta llegar al siguiente fracaso, en lugar de generar cada variable una por una.

### **Ejercicio 5**

In [13]:
random.seed(45)

resultados_posibles = set(range(2, 13))

def simular_lanzamientos_dados():
    resultados_obtenidos = set()
    lanzamientos = 0

    while resultados_obtenidos != resultados_posibles:
        dado_1 = random.randint(1, 6)
        dado_2 = random.randint(1, 6)
        suma = dado_1 + dado_2

        resultados_obtenidos.add(suma)
        lanzamientos += 1

    return lanzamientos

simulaciones = 100000
lanzamientos_necesarios = [simular_lanzamientos_dados() for _ in range(simulaciones)]

serie_lanzamientos = pd.Series(lanzamientos_necesarios)

resumen_dados = pd.DataFrame({
    "Métrica": [
        "Simulaciones realizadas",
        "Promedio de lanzamientos",
        "Mínimo de lanzamientos",
        "Máximo de lanzamientos",
        "Desviación estándar"
    ],
    "Valor": [
        simulaciones,
        round(serie_lanzamientos.mean(), 2),
        int(serie_lanzamientos.min()),
        int(serie_lanzamientos.max()),
        round(serie_lanzamientos.std(), 2)
    ]
})

resumen_dados

,Métrica,Valor
0,Simulaciones realizadas,100000.00
1,Promedio de lanzamientos,61.03
2,Mínimo de lanzamientos,11.00
3,Máximo de lanzamientos,413.00
4,Desviación estándar,35.81



Con la simulación realizada, el número esperado de lanzamientos necesarios para obtener todos los resultados posibles de 2 a 12 es aproximadamente 61.03 lanzamientos. 

### **Ejercicio 6**

Para estimar la suma de los valores de los elementos distintos, se pueden seleccionar posiciones aleatorias de la lista. Cada vez que se selecciona una posición, se observa qué elemento aparece y se revisa si ya había sido tomado en cuenta antes. Si el elemento no había aparecido, se agrega a un conjunto de elementos encontrados y se suma su valor. Si ya había aparecido, no se vuelve a sumar.
Si se repite este proceso muchas veces se obtiene una muestra de los elementos distintos de la lista. Con esa muestra se puede estimar la suma total, especialmente cuando la lista es demasiado grande como para recorrerla completa.

### **Ejercicio 7**

**a)**  
En el paso 2, la variable Y tiene una distribución geométrica con parámetro λ. Esto significa que Y representa cuántos pasos se avanzan hasta que ocurre un éxito. Su función de probabilidad es:

$$P(Y = k) = (1-\lambda)^{k-1}\lambda$$

para k = 1, 2, 3, ...

**b)**  
Lo que hace el algoritmo es ir avanzando por bloques. Primero genera un valor Y, que indica cuántas posiciones se deben avanzar. Ese avance se suma a S, entonces S queda como una posible posición donde podría ocurrir el evento. Después se genera otro número aleatorio para decidir si esa posición se acepta o no. Si se acepta, el algoritmo termina y toma X = S. Si no se acepta, vuelve a generar otro bloque y sigue avanzando.

**c)**  
El algoritmo es válido porque λ funciona como una cota para todas las tasas λn. Como cada λn es menor o igual que λ, entonces λn / λ siempre queda entre 0 y 1, por lo que se puede usar como probabilidad de aceptación. De esta manera, la distribución geométrica propone posibles posiciones y el paso de aceptación ajusta esas posiciones según la tasa real λn. Por eso, el valor final X termina teniendo las tasas de riesgo discretas {λn}.

### **Ejercicio 8**

La idea del algoritmo es generar primero un valor de X y luego decidir si ese valor se acepta o no. Esa decisión se hace usando la probabilidad P(Y = j | X = i), o sea, qué tan probable es que ocurra Y = j sabiendo que salió ese valor de X.

Si el valor no se acepta, simplemente se vuelve a generar otro valor de X y se repite el proceso. Entonces, los valores que tienen mayor relación con el evento Y = j tienen más probabilidad de ser aceptados.

Por eso, al final los valores aceptados no siguen la distribución original de X, sino la distribución de X bajo la condición de que Y = j. En otras palabras, el algoritmo va filtrando los valores de X hasta quedarse con los que corresponden a esa condición.

Por esa razón, el resultado final cumple que:

$$P(W = i) = P(X = i \mid Y = j)$$